Modules/_threadmodule.c

```diff
@@ -1242,13 +1276,28 @@ thread__set_sentinel(PyObject *self, PyObject *Py_UNUSED(ignored))
         Py_DECREF(lock);
         return NULL;
     }
-    tstate->on_delete_data = (void *) wr;
+
+    /* We can't use a tuple, because we can't deallocate the tuple without
+       a valid thread state. */
+    data = PyMem_NEW(PyObject *, 3);
+    if (data == NULL) {
+        Py_DECREF(lock);
+        Py_DECREF(wr);
+        return NULL;
+    }
+    data[0] = wr;
+    Py_INCREF(shutdown_locks_lock);
+    data[1] = shutdown_locks_lock;
+    Py_INCREF(shutdown_locks);
+    data[2] = shutdown_locks;
+
+    tstate->on_delete_data = (void *) data;
```

Comment:
_On_delete_data_ is no longer just a weak reference to a lock, so the comment in Include/cpython/pystate.h is out of date.